### ISSUE TITLE
Allow mining when specified too many GPUs

### ISSUE THREAD

**Issue by MoneroCrusher** [fireice-uk/xmr-stak#1882]
I run 12 GPU rigs in Linux and from time to time it happens that a GPU crashes and I have to take one out (maybe because a fan died or any other reason) and replace it at a later time.
Now when I start the miner with 11 GPUs it says GPU index 11 missing, and the miner only starts for CPU, ignoring the 11 GPUs that are in the system.
Instead of having to modify the config file, could you just let the 11 GPUs hash anyways and maybe make a small warning that 1 GPU is missing?

Another problem I have (Ubuntu 16.04): when a GPU crashes and I ctrl-c the miner, the miner just freezes indefinitely and I have to kill the machine. Is there a way to not kill the machine and still be able to reboot?

**Comment by Spudz76** [comment]
First part, probably a good idea.  I personally run a modified version that doesn't exit for much of any reason, the other place that bugs me is the HTTPd port clashing (should also be a 'soft' error, not panic and exit).  Do you run OpenCL or CUDA?

Second part, that is the driver hanging, you can pop a second terminal and reboot from there.  The kmod will not unload and none of the processes that had compute sessions open will revive or exit.  As it is in the driver we have no control over "gracefully" containing the external explosion in any app code here.

I had similar issues but once I revisited and overhauled my power cabling all issues went away.  I would imagine 12 GPU need some high quality powering (not just watts @ PSUs but also good thick cables and less than 160w per whip ~ single 6pin and a slot extension).  Also watch for insertion losses at any modular dumb connectors, mine like to melt even at 160w each (I solder 12+gauge wire direct to the internal rails now, bypass extra garbage connectors, have not had driver crashes since)

Point being the driver doesn't just lose the GPU normally, and would know where it went and recover if it could.  The only real thing that can make a GPU confuse everything is seizing up or falling off the bus, aka power problems or bad quality control on the slot extension (I had one out of ~24 with not enough solder or bridged contacts or other defects).  Nvidia drivers actually have a kernel message "GPU fell off the bus" and sometimes recovers successfully, however AMD does not.

**Comment by MoneroCrusher** [comment]
I run OpenCL/AMD.
The problem is when I open a second terminal and normally reboot it will just reboot indefinitely (last time I stopped measuring at the 1 hour mark).
So in that case I have to kill it, which surely isn't optimal for the fs..
```
echo 1 > /proc/sys/kernel/sysrq
echo s > /proc/sysrq-trigger
sleep 2
echo b > /proc/sysrq-trigger
```
That 160W is exactly what I draw per 6 pin slot, so maybe thats the problem. I'm using 94% platinum HP server PSUs with breakout boards and the black/yellow 6 pin cables. Haven't had a melt so far.

Would it somehow be possible to make a script for AMD GPUs to give that error message as well?

**Comment by Spudz76** [comment]
Yes, Systemd will hang unless it can kill everything, are you running under the systemd service file (available in PR's)?

No the driver does the cool stuff, AMD driver just isn't written to catch runtime disappearances.  Nvidia driver doesn't always catch it either, and the rest of the cards sometimes don't come back when one annoys the system.

Also SATA->slot power is not enough, the flat SATA power connectors have a single face of contact and can only do about 35w constant (slot eats 70-80w depending what the fans are doing).  If SATA->slot power is involved convert to direct molex 4-pin (best) or 6-pin (regular pcie power).  Usually they quit transferring current before they get hot enough to show melting.
